backend/src/v2/driver/driver.go

```diff
@@ -2006,7 +2028,253 @@ func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.A
+
+		// Base cases for handling the task that actually maps to the task that
+		// created the artifact/parameter we are searching for
+		// Base case 1: currentTask is a ContainerExecution that we can load the values off of
+		if *currentTask.GetExecution().Type == "system.ContainerExecution" {
+			glog.V(4).Infof("currentTask, %v, is a ContainerExecution", currentTaskName)
+			if isArtifact {
+				outputArtifacts, err := cfg.mlmd.GetOutputArtifactsByExecutionId(cfg.ctx, currentTask.GetID())
+				if err != nil {
+					return nil, nil, err
+				}
+				glog.V(4).Infof("outputArtifacts: %#v", outputArtifacts)
+				glog.V(4).Infof("outputKey: %v", workingOutputKey)
+				runtimeArtifact, err := outputArtifacts[workingOutputKey].ToRuntimeArtifact()
+				if err != nil {
+					return nil, nil, cfg.err(err)
+				}
+				glog.V(4).Infof("runtimeArtifact: %v", runtimeArtifact)
+				parallelForArtifactList = append(parallelForArtifactList, runtimeArtifact)
+			} else {
+				_, outputParameters, err := currentTask.GetParameters()
+				glog.V(4).Infof("outputParameters: %v", outputParameters)
+				if err != nil {
+					return nil, nil, cfg.err(err)
+				}
+				parallelForParameterList = append(parallelForParameterList, outputParameters[workingOutputKey])
+			}
+			continue
+
+			// Base case 2: currentTask is a DAGExecution within a loop but is NOT a ParallelFor
+			// Head DAG
+		} else if iterations == nil {
+			tempSubTaskName := workingSubTaskName
+			if iterationIndex != nil {
+				// handle for parallel iteration dag
+				tempSubTaskName = metadata.GetParallelForTaskName(tempSubTaskName, iterationIndex.GetIntValue())
+				glog.V(4).Infof("subTaskIterationName: %v", tempSubTaskName)
+			}
+			glog.V(4).Infof("tempSubTaskName: %v", tempSubTaskName)
+			resolveTaskNameQueue = append(resolveTaskNameQueue, tempSubTaskName)
+			continue
+		}
+
+		for i := range iterations.GetIntValue() {
+			loopName := metadata.UpdateTaskNameWithDagID(currentTask.TaskName(), currentTask.GetID())
+			loopIterationName := metadata.GetParallelForTaskName(loopName, i)
+			glog.V(4).Infof("loopIterationName: %v", loopIterationName)
+			resolveTaskNameQueue = append(resolveTaskNameQueue, loopIterationName)
+		}
+	}
+
+	outputParameterList := &structpb.Value{
+		Kind: &structpb.Value_ListValue{
+			ListValue: &structpb.ListValue{
+				Values: parallelForParameterList,
+			},
+		},
+	}
+	outputArtifactList := &pipelinespec.ArtifactList{
+		Artifacts: parallelForArtifactList,
+	}
+	glog.V(4).Infof("outputParameterList: %v", outputParameterList)
+	glog.V(4).Infof("outputArtifactList: %v", outputArtifactList)
+	return outputParameterList, outputArtifactList, nil
+}
+
+// Helper for determining whether to update the ProducerSubTask / Output[Artifact|Parameter]Key
+func InspectProducerTask(parentTask *metadata.Execution, tasks map[string]*metadata.Execution, subTaskName string, outputKey string, isArtifact bool) (string, string, error) {
```

Comment:
I recommend putting these parameters on separate lines to reduce the line length of the function signature. It may also help add a more elaborate doc string and maybe even named returns in the signature.